obspy/core/stream.py

```diff
@@ -3101,6 +3134,205 @@ def _dummy_stream_from_string(s):
+        ids_ = set([tr.id.rsplit(".", 1)[0] for tr in self])
+        all_channels = {}
+        # work can be separated by net.sta.loc, so iterate over each
+        for id_ in ids_:
+            net, sta, loc = id_.split(".")
+            channels = {}
+            st_ = self.select(network=net, station=sta, location=loc)
+            # for each individual channel collect earliest start time and
+            # latest endtime
+            for tr in st_:
+                cha = tr.stats.channel
+                cha_common = cha and cha[:-1] or None
+                cha_common_dict = channels.setdefault(cha_common, {})
+                cha_dict = cha_common_dict.setdefault(cha, {})
+                cha_dict["start"] = min(tr.stats.starttime.timestamp,
+                                        cha_dict.get("start", np.inf))
+                cha_dict["end"] = max(tr.stats.endtime.timestamp,
+                                      cha_dict.get("end", -np.inf))
+            # convert all timestamp objects back to UTCDateTime
+            for cha_common_dict in channels.values():
+                for cha_dict in cha_common_dict.values():
+                    cha_dict["start"] = UTCDateTime(cha_dict["start"])
+                    cha_dict["end"] = UTCDateTime(cha_dict["end"])
+            # now for every combination of common channels determine earliest
+            # common start time and latest common end time, as well as gap
+            # information in between
+            for cha_common, channels_ in channels.items():
+                if cha_common is None:
+                    cha_pattern = ""
+                else:
+                    cha_pattern = cha_common + "?"
+                st__ = self.select(network=net, station=sta, location=loc,
+                                   channel=cha_pattern)
+                start = max(
+                    [cha_dict_["start"] for cha_dict_ in channels_.values()])
+                end = min(
+                    [cha_dict_["end"] for cha_dict_ in channels_.values()])
+                gaps = st__.get_gaps()
+                all_channels[(net, sta, loc, cha_pattern)] = {
+                    "start": start, "end": end, "gaps": gaps,
+                    "channels": channels_}
+        return all_channels
+
+    def _trim_common_channels(self):
+        """
+        Trim all channels that have the same ID down to the component character
+        to the earliest common start time and latest common end time. Works in
+        place.
+        """
+        self._cleanup()
+        channel_infos = self._get_common_channels_info()
+        new_traces = []
+        for (net, sta, loc, cha_pattern), infos in channel_infos.items():
+            st = self.select(network=net, station=sta, location=loc,
+                             channel=cha_pattern)
+            st.trim(infos["start"], infos["end"])
+            for _, _, _, _, start_, end_, _, _ in infos["gaps"]:
+                st = st.cutout(start_, end_)
```

Comment:
One side effect of the cutout is that it changes the number of traces in the stream if gaps are present, which is a bit unexpected at the first glance. I assume it is no alternative to mask the data?

Reply:
Yeah, this might be unexpected for users, but in general relying on something like `stream[0]` in user code without using things like `stream.select(component=...)` and checking length of that output (which I see really often) is just plain bad practice.

Plus, masking data might lead to even much more obscure problems to unsuspecting users, with all the implications that can come with masked arrays and signal processing operations that work on the underlying numpy arrays. So I think this is by a wide margin the lesser evil.